### PR TITLE
EES-5464 replace CSV download rewrite with middleware redirect

### PR DESCRIPTION
**Pull request by bennettstuart**
replaces CSV download rewrite with middleware redirect.  
A smallish refactor on how we've had middleware setup, it should be easier to chain middleware in the future.  
following the above, tests have been updated

### CODE REVIEW

**Review by mmyoungman — APPROVED** [review]

File: src/explore-education-statistics-frontend/src/middleware/pages/redirectPages.ts [hunk 1]
```diff
@@ -19,66 +19,68 @@ const redirectPaths = {
   publications: '/find-statistics',
 };
 
-export default async function redirectPages(request: NextRequest) {
-  const { nextUrl } = request;
-  const decodedPathname = decodeURIComponent(request.nextUrl.pathname);
+export default function redirectPages(middleware: NextMiddleware) {
+  return async (request: NextRequest, event: NextFetchEvent) => {
```
Comment (ntsim) [inline src/explore-education-statistics-frontend/src/middleware/pages/redirectPages.ts:23]: Think we could make this a bit flatter and more Express-like by not making this a curried function. 

We could change the middleware function signature to:

```ts
export default async function redirectPages(
  request: NextRequest,
  event: NextFetchEvent,
  middleware: NextMiddleware,
) {
  // Rest of code...
}
```

As we don't need to return a function anymore, we avoid the extra indentation and don't need to change the entire function.

We just need to make some tweaks to `chain.ts`:

```ts
import { NextMiddlewareResult } from 'next/dist/server/web/types';
import {
  NextFetchEvent,
  NextMiddleware,
  NextRequest,
  NextResponse,
} from 'next/server';

export type ChainedMiddleware = (
  request: NextRequest,
  event: NextFetchEvent,
  next: NextMiddleware,
) => NextMiddlewareResult | Promise<NextMiddlewareResult>;

export default function chain(
  middlewares: ChainedMiddleware[],
): NextMiddleware {
  const [current, ...remaining] = middlewares;

  if (current) {
    const next = chain(remaining);
    return (request, event) => current(request, event, next);
  }

  return () => NextResponse.next();
}
```

Note we don't need to expose an `index` parameter as we can just take the first element from the array.

This also needs a slight tweak to the middleware test function:

```ts
export default function middlewareTestRun(
  middleware: ChainedMiddleware,
  page: string,
) {
  const request = new NextRequest(new Request(page));

  const fetchEvent = new NextFetchEvent({
    page,
    request,
  });

  return middleware(request, fetchEvent, () => NextResponse.next());
}
```

**Review by ntsim — CHANGES_REQUESTED** [review]

File: src/explore-education-statistics-frontend/package.json [hunk 2]
```diff
@@ -25,6 +25,7 @@
     "memoizee": "^0.4.14",
     "next": "^14.2.2",
     "next-sitemap": "^4.2.3",
+    "node-match-path": "^0.6.3",
```
Comment (ntsim) [inline src/explore-education-statistics-frontend/package.json:28]: As per other comments - don't need this

File: src/explore-education-statistics-frontend/src/middleware/pages/__tests__/rewriteDataCatalogue.test.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { NextResponse } from 'next/server';
+import rewriteDataCatalogueDownload from '../rewriteDataCatalogue';
+import middlewareTestRun from './util/middlewareTetsRun';
+
+describe('rewriteDataCatalogue', () => {
+  const redirectSpy = jest.spyOn(NextResponse, 'redirect');
+  const nextSpy = jest.spyOn(NextResponse, 'next');
+
+  test('redirects data-catalogue csv downloads to file download endpoint', async () => {
```
Comment (ntsim) [inline src/explore-education-statistics-frontend/src/middleware/pages/__tests__/rewriteDataCatalogue.test.ts:9]: A reminder that this test needs to be updated appropriately in the context of the proposed `rewrites` middleware changes:

- naming of `describe`
- naming of test case - 'rewrite' not 'redirect'

File: src/explore-education-statistics-frontend/src/middleware/chain.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  NextResponse,
+} from 'next/server';
+
+export type ChainedMiddleware = (
+  request: NextRequest,
+  event: NextFetchEvent,
+  next: NextMiddleware,
+) => NextMiddlewareResult | Promise<NextMiddlewareResult>;
+
+export default function chain(
```
Comment (ntsim) [inline src/explore-education-statistics-frontend/src/middleware/chain.ts:15]: We should move this into `middleware/pages` as it's only relevant for page middleware

**Review by ntsim — APPROVED** [review]

**Pull request merged**